File: patterns/primer-mit-hablladin/rules-js/primer-mit-hablladin.js
const GLSM_OFF = "0";
const GLSM_ON = "1";
const GLSM_TIMED_ON = "2";
const GLSM_TIMED_BLINK = "3";

let glsmTimer = null;

rules.JSRule({
  name: "GLSM - Garage Lights State Machine Events Handler",
  triggers: [triggers.ItemStateUpdateTrigger("GLSM")],
  execute: (event) => {
    const state = items.getItem("GLSM").state;

    if (state === "NULL" || state === "UNDEF") {
      items.getItem("GarageLights").sendCommand("OFF");
      items.getItem("GLSM").postUpdate(GLSM_OFF);
    } else if (state === GLSM_OFF) {
      items.getItem("GarageLights").sendCommand("OFF");
      if (glsmTimer !== null) { glsmTimer.cancel(); glsmTimer = null; }
    } else if (state === GLSM_ON) {
      items.getItem("GarageLights").sendCommand("ON");
      if (glsmTimer !== null) { glsmTimer.cancel(); glsmTimer = null; }
    } else if (state === GLSM_TIMED_ON) {
      items.getItem("GarageLights").sendCommand("ON");
      glsmTimer = actions.ScriptExecution.createTimer(
        time.ZonedDateTime.now().plusMinutes(5),
        () => items.getItem("GLSM").postUpdate(GLSM_TIMED_BLINK)
      );
    } else if (state === GLSM_TIMED_BLINK) {
      items.getItem("GarageLights").sendCommand("OFF");
      items.getItem("GarageLights").sendCommand("ON");
      glsmTimer = actions.ScriptExecution.createTimer(
        time.ZonedDateTime.now().plusMinutes(1),
        () => items.getItem("GLSM").postUpdate(GLSM_OFF)
      );
    }
  }
});

File: patterns/primer-mit-hablladin/rules-python/primer_mit_hablladin.py
import threading

from openhab import rule, Registry
from openhab.triggers import ItemStateUpdateTrigger

GLSM_OFF = "0"
GLSM_ON = "1"
GLSM_TIMED_ON = "2"
GLSM_TIMED_BLINK = "3"

glsm_timer = None
glsm_lock = threading.Lock()


@rule(triggers=[ItemStateUpdateTrigger("GLSM")])
class GLSMGarageLightsStateMachineEventsHandler:
    def execute(self, module, input):
        global glsm_timer

        with glsm_lock:
            state = str(Registry.getItem("GLSM").state)

            if state in ("NULL", "UNDEF"):
                Registry.getItem("GarageLights").sendCommand("OFF")
                Registry.getItem("GLSM").postUpdate(GLSM_OFF)
            elif state == GLSM_OFF:
                Registry.getItem("GarageLights").sendCommand("OFF")
                if glsm_timer is not None:
                    glsm_timer.cancel()
                    glsm_timer = None
            elif state == GLSM_ON:
                Registry.getItem("GarageLights").sendCommand("ON")
                if glsm_timer is not None:
                    glsm_timer.cancel()
                    glsm_timer = None
            elif state == GLSM_TIMED_ON:
                Registry.getItem("GarageLights").sendCommand("ON")

                def to_blink():
                    Registry.getItem("GLSM").postUpdate(GLSM_TIMED_BLINK)

                glsm_timer = threading.Timer(5 * 60, to_blink)
                glsm_timer.start()
            elif state == GLSM_TIMED_BLINK:
                Registry.getItem("GarageLights").sendCommand("OFF")
                Registry.getItem("GarageLights").sendCommand("ON")

                def to_off():
                    Registry.getItem("GLSM").postUpdate(GLSM_OFF)

                glsm_timer = threading.Timer(60, to_off)
                glsm_timer.start()
